lib/streamlit/commands/query_params.py

```diff
@@ -80,7 +92,50 @@ def set_query_params(**query_params: Any) -> None:
     ctx = get_script_run_ctx()
     if ctx is None:
         return
-    ctx.query_string = parse.urlencode(query_params, doseq=True)
+
     msg = ForwardMsg()
-    msg.page_info_changed.query_string = ctx.query_string
+    msg.page_info_changed.query_string = _ensure_no_embed_params(
+        query_params, ctx.query_string
+    )
     ctx.enqueue(msg)
+
+
+def _ensure_no_embed_params(
+    query_params: Dict[str, List[str]], query_string: str
+) -> str:
+    """Ensures there are no embed params set (raises StreamlitAPIException) if there is a try,
+    also makes sure old param values in query_string are preserved. Returns query_string : str."""
+    # Get query params dict without embed, embed_options params
+    query_params_without_embed = util.exclude_key_query_params(
+        query_params, keys_to_exclude=EMBED_QUERY_PARAMS_KEYS
+    )
+    if query_params != query_params_without_embed:
+        raise StreamlitAPIException(
+            "Query param embed and embed_options (case-insensitive) cannot be set using set_query_params method."
+        )
+
+    all_current_params = parse.parse_qs(query_string)
+    current_embed_params = parse.urlencode(
+        {
+            EMBED_QUERY_PARAM: [
+                param
+                for param in util.extract_key_query_params(
+                    all_current_params, param_key=EMBED_QUERY_PARAM
+                )
+            ],
+            EMBED_OPTIONS_QUERY_PARAM: [
+                param
+                for param in util.extract_key_query_params(
+                    all_current_params, param_key=EMBED_OPTIONS_QUERY_PARAM
+                )
+            ],
+        },
+        doseq=True,
+    )
+    query_string = parse.urlencode(query_params, doseq=True)
+
+    if query_string:
+        return (
+            f"{query_string}{'&' if current_embed_params else ''}{current_embed_params}"
```

Comment:
nit: you could do:

```
separator = '&' if current_embed_params else ''
separator.join([query_string, current_embed_params]) 
```

Reply:
It's updated 👍 Many thanks for all the reviews 🙇